commit ab7bf0476a3b09229c0f152ff1786b57b4942691
Author: bertin <macportsuser@macports.org>
Date:   Fri May 23 00:02:10 2025 +0200

    /opt/local/site-ports/sysutils/MacPorts/files/patch-cregistry-verbose+forced-vacuuming.diff

diff --git a/src/cregistry/registry.c b/src/cregistry/registry.c
index a778e2daa..0d597f8ec 100644
--- a/src/cregistry/registry.c
+++ b/src/cregistry/registry.c
@@ -499,6 +499,8 @@ int reg_rollback(reg_registry* reg, reg_error* errPtr) {
     }
 }
 
+int force_vacuum = 0;
+
 /**
  * Runs VACUUM (compact/defragment) on the given db file.
  * Works on a path rather than an open db pointer because you can't vacuum an
@@ -526,6 +528,7 @@ int reg_vacuum(char *db_path) {
 
     sqlite3_busy_timeout(db, 25);
 
+    fprintf(stderr, "Determining registry stats ..."); fflush(stderr);
     /* Get db size & size of unused space */
     if (sqlite3_prepare_v2(db, "SELECT page_count * page_size AS dbsize,"
             " freelist_count * page_size AS freesize FROM pragma_page_count(),"
@@ -541,21 +544,45 @@ int reg_vacuum(char *db_path) {
     if (stmt) {
         sqlite3_finalize(stmt);
     }
+    if (result) {
+        fprintf(stderr, " size=%gMB, free space = %gMB (%g%%)\n",
+            dbsize / 1.048576e06,
+            freesize / 1.048576e06, ((double)freesize / (double)dbsize) * 100);
+    } else {
+        fprintf(stderr, " failed (%s)\n", sqlite3_errmsg(db));
+    }
 
     /* Don't vacuum unless free space is at least 1 MB and also at
        least 1% of the total db size. */
-    if (result && freesize >= 1000000 && dbsize > 0 && ((double)freesize / (double)dbsize) >= 0.01) {
+    if (result && (force_vacuum || (freesize >= 1048576 && dbsize > 0 && ((double)freesize / (double)dbsize) >= 0.01))) {
         result = 0;
+        fprintf(stderr, "Vacuuming the registry"); fflush(stderr);
         if (sqlite3_prepare_v2(db, "VACUUM", -1, &stmt, NULL) == SQLITE_OK) {
-            sqlite3_step(stmt);
+            char *e;
+            fprintf(stderr, " ..."); fflush(stderr);
+            if (sqlite3_step(stmt) != SQLITE_OK && (e = sqlite3_errmsg(db))) {
+                fprintf(stderr, " (step: %s)"); fflush(stderr);
+            }
+            fprintf(stderr, " ..."); fflush(stderr);
             r = sqlite3_reset(stmt);
             if (r == SQLITE_OK) {
                 result = 1;
+            } else if (e = sqlite3_errmsg(db)) {
+                fprintf(stderr, " (reset: %s)"); fflush(stderr);
+            }
+        } else {
+            const char *e = sqlite3_errmsg(db);
+            if (e) {
+                fprintf(stderr, " (prepare: %s)", e);
             }
         }
         if (stmt) {
             sqlite3_finalize(stmt);
+            fprintf(stderr, result ? " done\n" : " failed\n");
+        } else {
+            fprintf(stderr, " failed\n");
         }
+        force_vacuum = 0;
     }
 
     sqlite3_close(db);
diff --git a/src/registry2.0/registry.c b/src/registry2.0/registry.c
index e9a6e1a64..b013b2e48 100644
--- a/src/registry2.0/registry.c
+++ b/src/registry2.0/registry.c
@@ -398,11 +398,13 @@ int metadata_cmd(ClientData clientData UNUSED, Tcl_Interp* interp, int objc,
 /* Allow setting the needs_vacuum flag from scripts. */
 int set_needs_vacuum_cmd(ClientData clientData UNUSED, Tcl_Interp* interp, int objc,
         Tcl_Obj* const objv[]) {
+    extern int force_vacuum;
     if (objc != 1) {
         Tcl_WrongNumArgs(interp, 1, objv, "cmd");
         return TCL_ERROR;
     }
     Tcl_SetAssocData(interp, "registry::needs_vacuum", NULL, (ClientData)1);
+    force_vacuum = 1;
     return TCL_OK;
 }
 
